### PR TITLE
[JENKINS-73849] Adding token and SSL verification disabling check for FIPS

### PR DESCRIPTION
**Pull request by PereBueno**
When in FIPs mode SSL and token verificaiton can not be disabled.
This change adds a form validation that will warn the user, fails with an `IllegalArgumentException` if trying to save regardless the warning and checks values on `readResolve`, in case config file was manually modified.

### Testing done
Added 3 tests and checked manually.

```[tasklist]
### Submitter checklist
- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [ ] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

File: src/test/java/org/jenkinsci/plugins/oic/OicSecurityRealmFipsTest.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        assertThat("Validation is ok", response.kind, is(FormValidation.Kind.OK));
+    }
+
+    // This test is not strictly needed as per
+    // https://github.com/jenkinsci/jep/blob/master/jep/237/README.adoc#backwards-compatibility
+    // Just adding it to cover corner cases
+    @Test
+    @LocalData
+    public void failsOnMigrationTest() {
+        assertTrue("We should get a ReactorException, startup failed", j.getError() instanceof ReactorException);
```
Comment (fcojfernandez) [inline src/test/java/org/jenkinsci/plugins/oic/OicSecurityRealmFipsTest.java]: I'm not fan of ussing `assertTrue` or `assertFalse` with non boolean fields as you miss information. If you use `assertThat` with https://hamcrest.org/JavaHamcrest/javadoc/1.3/org/hamcrest/core/IsInstanceOf.html you will know the actual value that is not matching the expected one

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java [hunk 2]
```diff
@@ -633,6 +647,9 @@ public void setPkceEnabled(boolean pkceEnabled) {
 
     @DataBoundSetter
     public void setDisableTokenVerification(boolean disableTokenVerification) {
+        if (FIPS140.useCompliantAlgorithms() && disableTokenVerification) {
+            throw new IllegalArgumentException(Messages.OicSecurityRealm_DisableTokenVerificationFipsMode());
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java]: throw a [FormException](https://javadoc.jenkins.io/hudson/model/Descriptor.FormException.html) ?

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java [hunk 3]
```diff
@@ -298,13 +299,26 @@ public OicSecurityRealm(
             throws IOException {
         // Needed in DataBoundSetter
         this.disableSslVerification = Util.fixNull(disableSslVerification, Boolean.FALSE);
+        if (FIPS140.useCompliantAlgorithms() && this.disableSslVerification) {
+            throw new IllegalArgumentException(Messages.OicSecurityRealm_DisableSslVerificationFipsMode());
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java]: throw a [FormException](https://javadoc.jenkins.io/hudson/model/Descriptor.FormException.html) ?

File: src/main/java/org/jenkinsci/plugins/oic/OicServerManualConfiguration.java [hunk 4]
```diff
@@ -262,6 +263,15 @@ public FormValidation doCheckScopes(@QueryParameter String value) {
...
 
+        @POST
+        public FormValidation doCheckIssuer(@QueryParameter String issuer) {
+            Jenkins.get().checkPermission(Jenkins.ADMINISTER);
+            if (Util.fixEmptyAndTrim(issuer) == null) {
+                return FormValidation.warning(Messages.OicSecurityRealm_IssuerRequired());
+            }
+            return FormValidation.ok();
+        }
+
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/oic/OicServerManualConfiguration.java]: not sure why this is here?

File: src/main/resources/org/jenkinsci/plugins/oic/OicSecurityRealm/help-clientSecret.html [hunk 5]
```diff
@@ -1,4 +1,5 @@
 <div>
-	The Client secret used by jenkins to authenticate itself to the openid connect provider.
-	Special value "none" can be used to disable client secret (only with client_secret_post authentication).
+	The Subject Idendifier Type to use as part of the flow with the openid connect provider.
+	This must be supported by the IdP.
+	Valid values are <code>public</code> and <code>pairwise</code>.
```
Comment (jtnord) [inline src/main/resources/org/jenkinsci/plugins/oic/OicSecurityRealm/help-clientSecret.html]: I think this PR is out of date with the target branch (or incorrect merge has been performed?)

File: src/test/java/org/jenkinsci/plugins/oic/OicSecurityRealmFipsTest.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    @Test
+    @WithoutJenkins
+    public void settingNonCompliantValuesNotAllowedTest() throws IOException {
+        OicSecurityRealm realm = new OicSecurityRealm("clientId", Secret.fromString("secret"), null, false);
+        IllegalArgumentException ex = assertThrows(
+                IllegalArgumentException.class,
+                () -> new OicSecurityRealm("clientId", Secret.fromString("secret"), null, true));
+        assertThat(
+                "Exception contains the reason",
+                ex.getLocalizedMessage(),
```
Comment (jtnord) [inline src/test/java/org/jenkinsci/plugins/oic/OicSecurityRealmFipsTest.java]: this will be unreliable if run with a different Locale?

File: src/test/java/org/jenkinsci/plugins/oic/OicSecurityRealmFipsTest.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    @Test
+    @WithoutJenkins
+    public void validationWarnsOfInvalidValuesTest() {
+        OicSecurityRealm.DescriptorImpl descriptor = new OicSecurityRealm.DescriptorImpl();
+        FormValidation response = descriptor.doCheckDisableSslVerification(true);
+        assertThat("Shows an error", response.kind, is(FormValidation.Kind.ERROR));
+        assertThat(
+                "States SSL verification can not be disabled",
+                response.getMessage(),
+                containsString("SSL verification can not be disabled"));
```
Comment (jtnord) [inline src/test/java/org/jenkinsci/plugins/oic/OicSecurityRealmFipsTest.java]: NIT: these can be combined, see https://github.com/jenkinsci/oic-auth-plugin/blob/c7c0c06e8af5510f3359c0ec57ed0bbca6bb0529/src/test/java/org/jenkinsci/plugins/oic/OicServerWellKnownConfigurationTest.java#L49-L59

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java (hunk 3 above)
Comment (fcojfernandez) [reply src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java]: agree, as in https://github.com/jenkinsci/oic-auth-plugin/pull/418

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java [hunk 8]
```diff
@@ -298,13 +299,26 @@ public OicSecurityRealm(
...
     protected Object readResolve() throws ObjectStreamException {
+        // Fail if migrating to a FIPS non-compliant config
+        if (FIPS140.useCompliantAlgorithms()) {
+            if (isDisableSslVerification()) {
+                throw new IllegalArgumentException(Messages.OicSecurityRealm_DisableSslVerificationFipsMode());
+            }
+            if (isDisableTokenVerification()) {
+                throw new IllegalArgumentException(Messages.OicSecurityRealm_DisableTokenVerificationFipsMode());
+            }
+        }
```
Comment (fcojfernandez) [inline src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java]: what's the correct exception for this method? Should be FormException as well? I'm tempted to say `IllegalStateException`
```suggestion
        if (FIPS140.useCompliantAlgorithms()) {
            if (isDisableSslVerification()) {
                throw new IllegalStateException(Messages.OicSecurityRealm_DisableSslVerificationFipsMode());
            }
            if (isDisableTokenVerification()) {
                throw new IllegalStateException(Messages.OicSecurityRealm_DisableTokenVerificationFipsMode());
            }
        }
```

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java [hunk 9]
```diff
@@ -1356,4 +1390,41 @@ public Descriptor<OicServerConfiguration> getDefaultServerConfigurationType() {
...
+
+    private static TokenAuthMethod toSupportedAuthMode(ClientAuthenticationMethod auth) {
+        String value = auth.getValue();
+        for (TokenAuthMethod tam : TokenAuthMethod.values()) {
+            if (tam.toString().equals(value)) {
+                return tam;
+            }
+        }
+        return null;
+    }
```
Comment (fcojfernandez) [inline src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java]: is this used anywhere?

File: src/main/java/org/jenkinsci/plugins/oic/OicServerManualConfiguration.java (hunk 4 above)
Comment (PereBueno) [reply src/main/java/org/jenkinsci/plugins/oic/OicServerManualConfiguration.java]: No idea, I guess some merge problem, just removed (was duplicate)

File: src/test/java/org/jenkinsci/plugins/oic/OicSecurityRealmFipsTest.java (hunk 6 above)
Comment (PereBueno) [reply src/test/java/org/jenkinsci/plugins/oic/OicSecurityRealmFipsTest.java]: changed to getMessage

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java (hunk 8 above)
Comment (PereBueno) [reply src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java]: I guess both would be ok, just using it as it's the normally used exception (e.g.: https://github.com/jenkinsci/ldap-plugin/blob/master/src/main/java/jenkins/security/plugins/ldap/LDAPConfiguration.java#L195-L202)

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java (hunk 9 above)
Comment (PereBueno) [reply src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java]: Seems only `toSupportedAuthMode`, the other 2 methods seem not used.
Anyway, these changes come from a push in the `pac4j` branch, so I would say maybe better check if they're used there.

**Review by fcojfernandez — APPROVED** [review]

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java (hunk 8 above)
Comment (fcojfernandez) [reply src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java]: no strong opinion

**Comment by PereBueno** [comment]
Now in the real repo https://github.com/jenkinsci/oic-auth-plugin/pull/423
